### PR TITLE
p2p/pex: reuse hash.Hasher per addrbook for speed

### LINKED ISSUE #6508: p2p/pex: optimize addrbook.hash
## Summary

Make addrbook.hash faster

## Problem Definition

Everytime `addrbook.hash` is called, we create new hasher and hash the data.

https://github.com/tendermint/tendermint/blob/1e0efb675899a9217d289db02c3f226b54e036c4/p2p/pex/addrbook.go#L940

## Proposal

We can pre-create the hasher and use it to hash the data.

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by cuonglm**
By pre-creating the hasher, instead of creating new one everytime
addrbook.hash is called.

```
name             old time/op    new time/op    delta
AddrBook_hash-8     181ns ±13%      80ns ± 1%  -56.08%  (p=0.000 n=10+10)

name             old alloc/op   new alloc/op   delta
AddrBook_hash-8      216B ± 0%        8B ± 0%  -96.30%  (p=0.000 n=10+10)

name             old allocs/op  new allocs/op  delta
AddrBook_hash-8      2.00 ± 0%      1.00 ± 0%  -50.00%  (p=0.000 n=10+10)
```

Fixed #6508

### CODE REVIEW

**Comment by cuonglm** [comment]
cc @odeke-em

**Review by marbar3778 — APPROVED** [review]
Wow this is awesome!! LGTM

Could you please add a changelog entry?

**Review by odeke-em — APPROVED** [review]
LGTM, thank you @cuonglm! Please make the title say:

p2p/pex: make and reuse hash.Hasher for speed

**Comment by odeke-em** [comment]
Rather:

p2p/pex: reuse hash.Hasher per addrbook for speed

**Review by cmwaters — APPROVED** [review]
Nice work. I think we should also backport these changes to 0.34

**Comment by cuonglm** [comment]
> Wow this is awesome!! LGTM
> 
> Could you please add a changelog entry?

Done.

File: p2p/pex/addrbook.go [hunk 1]
```diff
@@ -938,10 +941,7 @@ func groupKeyFor(na *p2p.NetAddress, routabilityStrict bool) string {
 }
 
 func (a *addrBook) hash(b []byte) ([]byte, error) {
```
Comment (alexanderbez) [inline p2p/pex/addrbook.go:943]: Is there concurrent access to this?

Comment (cuonglm) [reply p2p/pex/addrbook.go:943]: Yes, but it's guarded with a mutex by the caller.

**Review by alexanderbez — APPROVED** [review]

**Review by tychoish — APPROVED** [review]

**Comment by odeke-em** [comment]
Hey folks, please help us merge this fully approved PR otherwise it goes stale :-) Thank you.

**Comment by marbar3778** [comment]
> Hey folks, please help us merge this fully approved PR otherwise it goes stale :-) Thank you.

should be good to go now when the branch is updated to master

**Pull request merged**